store/layer_generators/retry_layer.go.tmpl

```diff
@@ -0,0 +1,101 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+// Code generated by "make store-layers"
+// DO NOT EDIT
+
+package store
+
+import (
+	"context"
+	
+	"github.com/lib/pq"
+	"github.com/mattermost/mattermost-server/v5/model"
+    "github.com/pkg/errors"
+	"github.com/go-sql-driver/mysql"
+)
+
+type {{.Name}} struct {
+	Store
+{{range $index, $element := .SubStores}}	{{$index}}Store {{$index}}Store
+{{end}}
+}
+
+{{range $index, $element := .SubStores}}func (s *{{$.Name}}) {{$index}}() {{$index}}Store {
+	return s.{{$index}}Store
+}
+
+{{end}}
+
+{{range $index, $element := .SubStores}}type {{$.Name}}{{$index}}Store struct {
+	{{$index}}Store
+	Root *{{$.Name}}
+}
+
+{{end}}
+
+func isRepeatableError(err error) bool {
+	var pqErr *pq.Error
+	var mysqlErr *mysql.MySQLError
+	switch {
+	case errors.As(errors.Cause(err), &pqErr):
+		if pqErr.Code == "40001" || pqErr.Code == "40P01" {
+			return true
+		}
+	case errors.As(errors.Cause(err), &mysqlErr):
+		if mysqlErr.Number == mySQLDeadlockCode {
+			return true
+		}
+	}
+	return false
+}
+
+{{range $substoreName, $substore := .SubStores}}
+{{range $index, $element := $substore.Methods}}
+func (s *{{$.Name}}{{$substoreName}}Store) {{$index}}({{$element.Params | joinParamsWithType}}) {{$element.Results | joinResultsForSignature}} {
+    {{if $element.Results | len | eq 0}}
+        s.{{$substoreName}}Store.{{$index}}({{$element.Params | joinParams}})
+    {{else}}
+        {{if $element.Results | errorPresent}}
+            {{if $element.Results | isAppError}}
+                return s.{{$substoreName}}Store.{{$index}}({{$element.Params | joinParams}})
+            {{else}}
+                tries := 0
+                for {
+                    {{$element.Results | genResultsVars}} := s.{{$substoreName}}Store.{{$index}}({{$element.Params | joinParams}})
+                    if {{$element.Results | errorVar}} == nil {
+                        return {{$element.Results | genResultsVars}}
+                    }
+                    if !isRepeatableError({{$element.Results | errorVar}}) {
+                        return {{$element.Results | genResultsVars}}
+                    }
+                    tries++
+                    if tries > 3 {
```

Comment:
If this is to mitigate deadlocks should we consider adding a small time (sleep?) before attempting again? This would help whatever is holding the lock finish doing its work.

Reply:
I'm ok introducing a tiny sleep, but normally, when you kill a deadlock transaction, you kill one of the offending transactions, so the other one is going to get the free lock and is going to finish it, and the retry one is going to succeed. (That is not necesarily alwasy true, so adding a random small amount of time there can help). What do you think @agnivade ?

Reply:
Not sure about the sleep part. My original retry function also doesn't have the sleep. I would expect this to work too. There is an extremely low probability that a valid transaction fails 3 times consecutively, but will pass if there's a sleep in between. This sounds like an edge-case of an already edge-case. Might not be worth adding it.